txstratum/protocol.py

```diff
@@ -0,0 +1,345 @@
+                self.miner_address_str = address
+            except InvalidAddress:
+                self.log.debug('Miner with invalid address', address=address)
+                self.send_error(msgid, self.INVALID_ADDRESS)
+                self.transport.close()
+                return
+        self._subscribed = True
+        self.send_result('ok', msgid)
+        self.log.info('Miner subscribed', address=self.miner_address_str)
+        self.start_if_ready()
+
+    def method_authorize(self, params: Any, msgid: JSONRPCId) -> None:
+        """Handle authorize request from JSONRPC."""
+        self._authorized = True
+        self.log.info('Miner authorized')
+        self.start_if_ready()
+
+    def method_submit(self, params: Any, msgid: JSONRPCId) -> None:
+        """Handle submit request from JSONRPC.
+
+        Args:
+            params: a dict containing `job_id` and `nonce`
+        """
+        self.log.debug('handle submit', msgid=msgid, params=params)
+        if 'job_id' not in params or 'nonce' not in params:
+            return self.send_error(msgid, self.INVALID_PARAMS, {'params': params, 'required': ['job_id', 'nonce']})
+
+        try:
+            job_id = bytes.fromhex(params['job_id'])
+        except ValueError:
+            return self.send_error(msgid, self.INVALID_PARAMS, 'invalid job_id format')
+
+        try:
+            nonce = int(params['nonce'], 16)
+        except ValueError:
+            return self.send_error(msgid, self.INVALID_PARAMS, 'invalid nonce')
+
+        job = self.jobs.get(job_id, None)
+
+        if job is None:
+            # XXX Should we look into the job history? It may be necessary if the job was not marked as clean.
+            return self.send_error(msgid, self.JOB_NOT_FOUND)
+
+        if job is not self.current_job or job.submitted_at is not None:
+            return self.send_error(msgid, self.STALE_JOB, {
+                'current_job': self.current_job and self.current_job.uuid.hex(),
+                'job_id': job_id.hex()
+            })
+
+        if job.is_block:
+            obj = Block.create_from_struct(job.get_data())  # type: BaseTransaction
+        else:
+            obj = Transaction.create_from_struct(job.get_data())
+
+        obj.nonce = nonce
+        obj.update_hash()
+        if not obj.verify_pow(override_weight=job.share_weight):
+            self.log.error('Invalid share weight', uuid=job.uuid.hex(), share_weight=job.share_weight)
+            return self.send_error(msgid, self.INVALID_SOLUTION)
+
+        now = time.time()
+        self.completed_jobs += 1
+        job.submitted_at = now
+        self.last_submit_at = now
+        self._submitted_work.append(SubmittedWork(job.submitted_at, job.share_weight))
+        self.send_result(msgid, 'ok')
+
+        # XXX Is it a good rule when mining txs?
+        # Too many jobs too fast, increase difficulty out of caution (more than 10 submits within the last 10s)
```

Comment:
I'm not sure if we should do that.

For blocks, it shouldn't happen. We probably won't have powerful miners connected to this API that could solve blocks very fast and, even if we did, weight adjustment should take care of it.

For transactions, I think it does not make sense.

Reply:
Do you think we should use only blocks to estimate the miner's hash rate?

Reply:
Yes, as we've had the random mining time for transactions, we should only use blocks for hash rate estimation.

Reply:
@msbrogli I still think this should be removed.

Reply:
Done!

Reply:
The changes are in the last commit. Can you review it, please? @obiyankenobi @pedroferreira1 @jansegre

Reply:
Looks good to me.